src/braket/experimental/autoqasm/api.py

```diff
@@ -321,6 +379,62 @@ def _convert_program_as_subroutine(
         root_oqpy_program._add_subroutine(subroutine_name, subroutine_function_call.subroutine_decl)
 
 
+def _convert_gate(
+    f: Callable,
+    declaration_only: bool,
+    options: converter.ConversionOptions,
+    args: List[Any],
+    kwargs: Dict[str, Any],
+) -> Callable:
+    try:
+        # We must be inside an active conversion context in order to invoke a gate
+        program_conversion_context = aq_program.get_program_conversion_context()
+
+        # Wrap the function into an oqpy gate definition
+        wrapped_f, gate_args = _wrap_for_oqpy_gate(f, options)
+        gate_name = f.__name__
+
+        # Validate that the gate definition acts on at least one qubit
+        qubits = [
+            oqpy.Qubit(name, needs_declaration=False) for name, is_qubit in gate_args if is_qubit
+        ]
+        if not qubits:
+            raise errors.ParameterTypeError(
+                f'Gate definition "{gate_name}" has no arguments of type aq.Qubit. '
+                "Every gate definition must contain at least one qubit argument."
+            )
+
+        if not declaration_only:
+            # Process the gate definition
+            angles = [oqpy.AngleVar(name=name) for name, is_qubit in gate_args if not is_qubit]
+            with program_conversion_context.gate_definition(gate_name, qubits, angles):
+                # TODO - enforce that nothing gets added to the program inside here except gates
+                wrapped_f(qubits, *angles)
+
+            # Add the gate definition to the root-level program if necessary
+            root_oqpy_program = program_conversion_context.oqpy_program_stack[0]
+            if gate_name not in root_oqpy_program.gates:
```

Comment:
What happens if you decorate two different functions with the same name? Can you add a couple tests for it?

Reply:
Adding a couple of tests. Basically I think the Python semantics should be respected here, the latest-defined function will be the one that is used.